---
title: "Exploratory Subsampled 10K Kraken2"
author: "Flora Bioworks"
date: "`r format(Sys.time(), '%a %B %d %X %Z %Y')`"
header-includes:
- \usepackage{float}
- \floatplacement{figure}{H}  #make every figure with caption = h
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
geometry: paperheight=11.7in,paperwidth=8.3in,margin=1in
output:
  html_document:
    code_folding: hide
    df_print: paged
    toc: true
    number_sections: true
    toc_depth: 6
    toc_float: true
  pdf_document: 
    keep_tex: no
    latex_engine: xelatex
    toc: yes
    toc_depth: 4
    fig_caption: yes
    includes:
      in_header: header.tex
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath("/media/justincsing/ExtraDrive1/Documents2/Roest_Lab/Github/microbiome_OJS")) 
```

# Dependencies
```{r message=FALSE, warning=FALSE}
library(reticulate)
library(data.table)
library(dplyr)
library(ggplot2)
library(dendextend)
library(phyloseq)
```

# Helper Functions
```{r}
do_lm <- function(dt_sub, grouped_var, value_col="value", group_col="ACVD", group_ctrl="0", group_treat="1"){
  
  # print(sprintf("Working on: %s", grouped_var[[1]]))
  
  ctrl <- dt_sub[[value_col]][dt_sub[[group_col]]==group_ctrl]
  treat <- dt_sub[[value_col]][dt_sub[[group_col]]==group_treat]
  
  # print(sprintf("nobs ctrl: %s, mean ctrl: %s | nobs treat: %s, mean treat: %s", length(ctrl), round(mean(ctrl, na.rm = T), 4), length(treat), round(mean(treat, na.rm = T), 4)))
  if ( (length(ctrl) > 1 & length(treat) > 1) & !all(is.na(ctrl)) & !all(is.na(treat)) ){
    lm.out <- lm(value ~ as.factor(ACVD), dt_sub)
    # interactions::interact_plot(lm.out, pred = value, modx = ACVD, plot.points = TRUE)
    # summary(lm.out)
    lm.out.df <- broom::tidy(lm.out)
    dt_sub$term <- lm.out.df$term[lm.out.df$term=="as.factor(ACVD)1"]
    dt_sub$pvalue <- lm.out.df$p.value[lm.out.df$term=="as.factor(ACVD)1"]
    dt_sub$foldchange <- lm.out.df$estimate[lm.out.df$term=="as.factor(ACVD)1"]
  } else {
    print(sprintf("WARN: Not Enough Observations to perform linear regression Returning NA"))
    dt_sub$term <- NA
    dt_sub$pvalue <- NA
    dt_sub$foldchange <- NA
  }
  return(dt_sub)
}
bin_age <- function(age){
  age <- as.numeric(age)
  if (is.na(age)){
    return(NA)
  }
  if (age >0 & age <10){
    return("0-10")
  } else if (age >=10 & age <20){
    return("10-20")
  } else if (age >=20 & age <30){
    return("20-30")
  } else if (age >=30 & age <40){
    return("30-40")
  } else if (age >=40 & age <50){
    return("40-50")
  } else if (age >=50 & age <60){
    return("50-60")
  } else if (age >=60 & age <70){
    return("60-70")
  } else {
    return("70+")
  }
}
```


# Files
```{r}
pickled_x <- "./workflow/PE_subsampling_experiments/insilico_depth_experiment/kraken2_PE_10K/notebooks/PE_10K_Sex/prepped_data/X.pk"
pickled_y <- "./workflow/PE_subsampling_experiments/insilico_depth_experiment/kraken2_PE_10K/notebooks/PE_10K_Sex/prepped_data/y.pk"
pickled_meta_data_mat <- "./workflow/PE_subsampling_experiments/insilico_depth_experiment/kraken2_PE_10K/notebooks/PE_10K_Sex/prepped_data/meta_data_mat.pk"
file_feat_meta_csv <- "./workflow/PE_subsampling_experiments/insilico_depth_experiment/kraken2_PE_10K/notebooks/PE_10K_Sex/prepped_data/feat_meta.csv"
file_metadata_samples_keep  <- "./workflow/PE_subsampling_experiments/insilico_depth_experiment/kraken2_PE_10K/notebooks/PE_10K_Sex/prepped_data/metadata_samples_keep.csv"
file_taxon_db <- "./data/taxonomy_db.csv"
```

# Load Data
```{r}
reticulate::source_python("./src/util/pickles_parser.py")
X <- read_pickle_file(pickled_x)
y <- read_pickle_file(pickled_y)
meta_data_mat <- read_pickle_file(pickled_meta_data_mat)
feat_meta_csv <- data.table::fread(file_feat_meta_csv, select=2L, header=T)
metadata_samples_keep <- data.table::fread(file_metadata_samples_keep)
taxonomy_db <- data.table::fread(file_taxon_db)
taxonomy_phylum_db <- taxonomy_db[ rank=="phylum"]
```

# Differential Expression Linear Model
```{r}
# Transpose and convert to matrix
X_t <- as.matrix(t(X))
colnames(X_t) <- metadata_samples_keep$RunID
rownames(X_t) <- feat_meta_csv$feature
# Create Sample Annotation mapping
run_id_to_acvd <- metadata_samples_keep[, .(RunID, ACVD=`ACVD status`, Gender, height=`Height (cm)`, bmi=`Body Mass Index (BMI)`)]
run_id_to_acvd$age_group <- unlist(lapply(metadata_samples_keep$`Age (year)`, bin_age))
run_id_to_acvd %>%
  head() %>%
  knitr::kable()
# Prepare long dt with sample annotation
X_t %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var="feature") %>%
  tidyr::pivot_longer(cols=colnames(X_t)) %>%
  dplyr::mutate(value=log2(value+1)) %>%
  merge(run_id_to_acvd, by.x="name", by.y="RunID") -> lm_prep
# Run linear model
lm_prep %>%
  dplyr::group_by(feature) %>%
  dplyr::group_map(do_lm, .keep=TRUE) %>%
  data.table::rbindlist() -> lm_dt

lm_dt %>%
  dplyr::select(feature, term, pvalue, foldchange) %>%
  unique() -> lm_dt

# remove nas
lm_dt <- lm_dt[ !is.na(lm_dt$pvalue), ]

hist(lm_dt$pvalue, )

# Multiple Testing correction
lm_dt$fdr <- p.adjust(lm_dt$pvalue, method="fdr")

hist(lm_dt$fdr)

lm_dt %>%
  dplyr::filter( fdr < 0.01 ) %>%
  dplyr::arrange( fdr) %>%
  knitr::kable()

lm_dt %>%
  dplyr::filter( fdr < 0.01 ) %>%
  dplyr::arrange( fdr) %>%
  dplyr::pull(feature) -> sig_feats
```

```{r}
# Filter Freq data for only the identified significant features identified by lm
X_t <- X_t[ rownames(X_t) %in% sig_feats, ]
rownames(X_t) <- paste0("OTU", 1:nrow(X_t))
# Create species feature to phylum mapping table
feat_meta_phylum <- as.matrix(plyr::mapvalues(sig_feats, from=taxonomy_phylum_db$query, to=taxonomy_phylum_db$name, warn_missing = FALSE))
feat_meta_phylum[feat_meta_phylum %in% feat_meta_csv$feature] <- "Unknown"
colnames(feat_meta_phylum) <- "Phylum"
rownames(feat_meta_phylum) <- paste0("OTU", 1:nrow(X_t))
feat_meta_phylum <- cbind(feat_meta_phylum, Species=sig_feats)
# Create phyloseq object
OTU <- phyloseq::otu_table(X_t, taxa_are_rows = T)
run_id_to_acvd <- as.data.frame(run_id_to_acvd)
rownames(run_id_to_acvd) <- run_id_to_acvd[,1]
run_id_to_acvd$ACVD <- as.factor(run_id_to_acvd$ACVD)
run_id_to_acvd$Gender <- as.factor(run_id_to_acvd$Gender)
ANN <- phyloseq::sample_data(run_id_to_acvd)
TAX <- phyloseq::tax_table(feat_meta_phylum)
physeq = phyloseq::phyloseq(OTU, TAX, ANN)
```

```{r}
phyloseq::plot_bar(physeq, fill = "Phylum") +
  geom_bar(aes(color = Phylum, fill = Phylum), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~ ACVD, scales = "free_x") +
  theme(panel.background = element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

```{r}
phyloseq::psmelt(physeq) %>%
  dplyr::filter(Species %in% sig_feats[1:6]) %>%
  ggplot(data = ., aes(x = ACVD, y = Abundance)) +
  geom_boxplot(outlier.shape  = NA) +
  geom_jitter(aes(color = Species), height = 0, width = .2, alpha=0.1) +
  labs(x = "", y = "Abundance\n") +
  facet_wrap(~ Species, scales = "free")

phyloseq::psmelt(physeq) %>%
  dplyr::filter(Species %in% sig_feats[7:12]) %>%
  ggplot(data = ., aes(x = ACVD, y = Abundance)) +
  geom_boxplot(outlier.shape  = NA) +
  geom_jitter(aes(color = Species), height = 0, width = .2, alpha=0.1) +
  labs(x = "", y = "Abundance\n") +
  facet_wrap(~ Species, scales = "free")
```

# Hierarchical Clustering
```{r}
#Extract OTU table and compute BC
ps_rel_otu <- data.frame(phyloseq::otu_table(physeq))
ps_rel_otu <- t(ps_rel_otu)
bc_dist <- vegan::vegdist(ps_rel_otu, method = "bray")
as.matrix(bc_dist)[1:5, 1:5]

#Save as dendrogram
ward <- as.dendrogram(hclust(bc_dist, method = "ward.D2"))
#Provide color codes
meta <- data.frame(phyloseq::sample_data(physeq))
colorCode <- c(`0` = "red", `1` = "blue")
colorGender <- c(male = "darkgreen", female = "purple")
colorAge <- stats::setNames(RColorBrewer::brewer.pal(length(unique(meta$age_group)), "Set1"), unique(meta$age_group))
# Add ACVD status annotation
# par(mar = c(7,2,1,1))
ward %>% dendextend::set("labels", "") %>% plot
dendextend::colored_bars(colors = cbind(colorGender[meta$Gender][order.dendrogram(ward)], colorAge[meta$age_group][order.dendrogram(ward)], colorCode[meta$ACVD][order.dendrogram(ward)]), dend = ward, rowLabels = c("Gender", "Age_group", "ACVD"), y_shift = -0.5)
```

