# 16S rRNA Sequencing

&emsp; 16S rRNA sequencing, based on amplicon sequencing, targets and reads a region of the 16S rRNA gene found in all Bacteria and Archae - limiting the detection of only these microorganisms. There are other separate amplicon sequencing methods that can identify other microorganisms, i.e. ITS for fungi or 18S for protists. 16S is good for determining the microbial composition of a sample and the relative abundance of microbes. The reliable resolution is only genra specific; you cannot ascertain species or strains.

# Shotgun Metagenomic Sequencing

&emsp; This method fragments DNA into small pieces, which are sequenced and then stiched back together using using bioinformatic pipelines. Metagenenomics is good for determining potential functionality of a sample and the contributors. I.e. you can assign genes with a high level of certainty to species (sometimes on the strain level) and identify which  species might be responsible for specific pathways/enzymes/gene clusters expresion.

# Shotgun Metatranscriptomics Sequencing

&emsp; Similar to metagenomics, but focuses on mRNA. Metatranscriptomics is the best option for determining which pathways/enzymes are expressed, with high confidence, and which taxa are responsible; estimate the real activity of a specific taxa. In the field, people generally draw conclusions from 16S and metagenomics about the activity of certain taxa, however, high abundance and presence of a lot of genes for a specific species may not be informative or interpreted as such or vice versa. Furethermore, metatranscriptomics is still generally new, so there are not a lot of reliable pre-processing tools that can reliably assign enzymes to taxa, and some tools filter out a lot of information in the pre-processing steps that may bias the results.

# Comparisons

|                                                   | 16S rRNA Sequencing                                                                                            | Shotgun Metagenomic Sequencing                                                                                       | Shotgun Metatranscriptomic Sequencing                                                                                                    |
| ------------------------------------------------- | -------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
| **Cost**                                          | ~$50 USD                                                                                                       | Starts at ~$150 USD, depends on sequencing depth                                                                     |                                                                                                                                          |
| **Functional profiling**                          | No (but 'predicted' functional profiling is possible)                                                          | Yes (but it only reveals information on functional potential)                                                        | Yes (best for assessing which pathways/enzymes are expressed and which taxa are responsiblel estimate the real activity of certain taxa) |
| **Taxonomic resolution: Genus, species, strain?** | Bacterial genus (sometimes species); dependent on region(s) targeted                                           | Bacterial species (sometimes strains and single nucleotide variants, if sequencing is deep enough)                   | Same as Metagenomics                                                                                                                     |
| **Taxonomic coverage**                            | Bacteria and archaea                                                                                           | All taxa, including viruses                                                                                          | Same as Metagenomics                                                                                                                     |
| **Bioinformatics requirements**                   | Beginner to intermediate expertise                                                                             | Intermediate to advanced expertise                                                                                   | Intermediate to advanced expertise                                                                                                       |
| **Databases**                                     | Established, well-curated                                                                                      | Relatively new, still growing                                                                                        | Relatively new, still growing                                                                                                            |
| **Sensitivity to host DNA contamination**         | Low (but PCR success depends on the absence of inhibitors and the presence of a detectable microbiome)         | High , varies with sample type (but this can be mitigated by calibrating the sequencing depth)                       | Same as Metagenomics                                                                                                                     |
| **Bias**                                          | Medium to high (retrieved taxonomic composition is dependent on selected primers and targeted variable region) | Lower (while metagenomics is “untargeted”, experimental and analytical biases can be introduced at various stages) | Same as Metagenomics                                                                                                                     |

# See

* [https://blog.microbiomeinsights.com/](https://blog.microbiomeinsights.com/)
* [https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-019-0743-116s-rrna-sequencing-vs-shotgun-metagenomic-sequencing](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-019-0743-116s-rrna-sequencing-vs-shotgun-metagenomic-sequencing)
* [https://www.frontiersin.org/articles/10.3389/fgene.2019.00904/full](https://www.frontiersin.org/articles/10.3389/fgene.2019.00904/full)
* [https://www.nature.com/articles/s41564-017-0089-z](https://www.nature.com/articles/s41564-017-0089-z)
* [https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4839964/](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4839964/)